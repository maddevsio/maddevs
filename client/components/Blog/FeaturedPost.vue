<template>
  <nuxt-link :to="link" class="featured-post">
    <div class="row featured-post__wrapper">
      <div class="col-12 col-lg-6 featured-post__main">
        <h1 class="featured-post__title">{{ $prismic.asText(post.data.title) }}</h1>
        <p class="featured-post__paragraph">{{ getFirstParagraph() }}</p>
        <div class="featured-post__data d-flex justify-content-between">
          <post-author :document="post.data" />
          <div class="featured-post__meta">
            <post-tag v-if="post.tags.length" :tag="post.tags[0]" theme="dark" />
            <span class="created-at">{{ formattedDate }}</span>
          </div>
        </div>
      </div>
      <div class="col-12 col-lg-6 featured-post__image-wrapper">
        <img
          :src="post.data.featured_image.url"
          :alt="post.data.featured_image.alt"
          :width="post.data.featured_image.dimensions.width"
          :height="post.data.featured_image.dimensions.height"
          class="featured-post__image"
        />
      </div>
    </div>
  </nuxt-link>
</template>

<script>
import PostAuthor from '@/components/Blog/PostAuthor'
import PostTag from '@/components/Blog/PostTag'
import linkResolver from '@/plugins/link-resolver.js'

export default {
  name: 'FeaturedPost',
  components: {
    PostAuthor,
    PostTag,
  },

  props: {
    post: {
      type: Object,
      required: true,
    },
  },

  computed: {
    formattedDate() {
      const date = this ? this.post.data.date : ''

      return Intl.DateTimeFormat('en-US', { year: 'numeric', month: 'short', day: '2-digit' }).format(new Date(date))
    },

    link() {
      return linkResolver(this.post)
    },
  },

  methods: {
    getFirstParagraph() {
      const textLimit = 200
      const slices = this.post.data.body
      let firstParagraph = ''
      let haveFirstParagraph = false

      slices.forEach(slice => {
        if (!haveFirstParagraph && slice.slice_type === 'text') {
          slice.primary.text.forEach(block => {
            if (block.type === 'paragraph' && !haveFirstParagraph) {
              firstParagraph += block.text
              haveFirstParagraph = true
            }
          })
        }
      })

      const limitedText = firstParagraph.substr(0, textLimit)

      return firstParagraph.length > textLimit
        ? `${limitedText.substr(0, limitedText.lastIndexOf(' '))}...`
        : firstParagraph
    },
  },
}
</script>

<style scoped lang="scss">
@import '../../assets/styles/_vars';

.featured-post {
  color: $text-color--white-primary;
  text-decoration: none;
  display: block;

  &__wrapper {
    margin: 95px 0;
    align-items: center;
  }

  &__main {
    margin-top: 0;
  }

  &__title {
    font-size: 52px;
    line-height: 130%;
    margin-bottom: 15px;
    letter-spacing: -2px;
    font-family: 'Poppins-Medium', sans-serif;
  }

  &__paragraph {
    font-size: 17px;
    line-height: 28px;
    letter-spacing: -0.035em;
    margin-bottom: 44px;
    font-family: 'Inter', sans-serif;
    font-weight: 400;
  }

  &__data {
    display: flex;
    justify-content: space-between;

    /deep/ .blog-post__author-name {
      color: $text-color--white-primary;
    }

    /deep/ .blog-post__author-image {
      width: 40px;
      height: 40px;
    }
  }

  &__meta {
    display: flex;
    align-items: center;
    font-size: 13px;
    font-family: 'Inter', sans-serif;
    font-weight: 400;

    .created-at {
      margin-left: 24px;
      color: $text-color--grey;
    }
  }

  &__image-wrapper {
    text-align: right;
    margin-top: 0;

    img {
      max-width: 90%;
      height: auto;
      vertical-align: middle;
    }
  }
}

@media only screen and (max-width: 991px) {
  .featured-post {
    &__wrapper {
      margin-top: 0;
      margin-bottom: 36px;
    }

    &__main {
      padding: 0;
      margin-top: 0;
      order: 2;
    }

    &__paragraph {
      margin-bottom: 38px;
      font-size: 16px;
      line-height: 150%;
      letter-spacing: -0.02em;
    }

    &__data {
      display: block;

      /deep/ .blog-post__author {
        margin-bottom: 16px;
      }

      /deep/ .blog-post__author-image {
        width: 30px;
        height: 30px;
      }
    }

    &__title {
      font-size: 35px;
    }

    &__meta {
      justify-content: space-between;

      /deep/ .post-tag {
        padding: 8px 16px;
        line-height: 22px;
        margin-left: 0;
      }
    }

    &__image-wrapper {
      padding: 0;
      order: 1;
      margin-bottom: 17px;
      margin-top: 0;
      text-align: center;

      img {
        max-width: 100%;
      }
    }
  }
}
</style>
