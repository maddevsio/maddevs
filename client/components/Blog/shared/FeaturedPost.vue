<template>
  <NuxtLink
    :to="link"
    class="featured-post"
  >
    <div class="row featured-post__wrapper">
      <div class="col-12 col-lg-6 featured-post__main">
        <h1 class="featured-post__title">
          {{ $prismic.asText(post.data.title) }}
        </h1>
        <p class="featured-post__paragraph">
          {{ firstParagraph }}
        </p>
        <div class="featured-post__data d-flex justify-content-between">
          <PostAuthor
            v-if="author"
            :author="author"
          />
          <div class="featured-post__meta">
            <PostTag
              v-if="post.tags.length"
              :tag="post.tags[0]"
              theme="dark"
            />
            <span class="created-at">{{ formattedDate }}</span>
          </div>
        </div>
      </div>
      <div class="col-12 col-lg-6 featured-post__image-wrapper">
        <img
          :src="post.data.featured_image.url"
          :alt="post.data.featured_image.alt"
          :width="post.data.featured_image.dimensions.width"
          :height="post.data.featured_image.dimensions.height"
          class="featured-post__image"
        >
      </div>
    </div>
  </NuxtLink>
</template>

<script>
import PostAuthor from '@/components/Blog/shared/PostAuthor'
import PostTag from '@/components/Blog/shared/PostTag'
import linkResolver from '@/plugins/link-resolver'
import getFirstParagraph from '@/helpers/getFirstParagraph'

export default {
  name: 'FeaturedPost',
  components: {
    PostAuthor,
    PostTag,
  },

  props: {
    post: {
      type: Object,
      required: true,
    },

    author: {
      type: Object,
      required: true,
    },
  },

  computed: {
    formattedDate() {
      const { date } = this.post.data
      return Intl.DateTimeFormat('en-US', { year: 'numeric', month: 'short', day: '2-digit' }).format(new Date(date))
    },

    link() {
      return linkResolver(this.post)
    },

    firstParagraph() {
      const limit = 200
      const slices = this.post.data.body
      return getFirstParagraph(slices, limit)
    },
  },
}
</script>

<style scoped lang="scss">
@import '../../../assets/styles/_vars';

.featured-post {
  color: $text-color--white-primary;
  text-decoration: none;
  display: block;

  &__wrapper {
    margin: 95px 0;
    align-items: center;
  }

  &__main {
    margin-top: 0;
  }

  &__title {
    font-size: 52px;
    line-height: 130%;
    margin-bottom: 15px;
    letter-spacing: -2px;
    font-family: 'Poppins-Medium', sans-serif;
  }

  &__paragraph {
    font-size: 17px;
    line-height: 28px;
    letter-spacing: -0.035em;
    margin-bottom: 44px;
    font-family: 'Inter', sans-serif;
    font-weight: 400;
  }

  &__data {
    display: flex;
    justify-content: space-between;

    /deep/ .blog-post__author-name {
      color: $text-color--white-primary;
    }

    /deep/ .blog-post__author-image {
      width: 40px;
      height: 40px;
    }
  }

  &__meta {
    display: flex;
    align-items: center;
    font-size: 13px;
    font-family: 'Inter', sans-serif;
    font-weight: 400;

    .created-at {
      margin-left: 24px;
      color: $text-color--grey;
    }
  }

  &__image-wrapper {
    text-align: right;
    margin-top: 0;

    img {
      max-width: 90%;
      height: auto;
      vertical-align: middle;
    }
  }
}

@media only screen and (max-width: 991px) {
  .featured-post {
    &__wrapper {
      margin-top: 0;
      margin-bottom: 36px;
    }

    &__main {
      padding: 0;
      margin-top: 0;
      order: 2;
    }

    &__paragraph {
      margin-bottom: 38px;
      font-size: 16px;
      line-height: 150%;
      letter-spacing: -0.02em;
    }

    &__data {
      display: block;

      /deep/ .blog-post__author {
        margin-bottom: 16px;
      }

      /deep/ .blog-post__author-image {
        width: 30px;
        height: 30px;
      }
    }

    &__title {
      font-size: 35px;
    }

    &__meta {
      justify-content: space-between;

      /deep/ .post-tag {
        padding: 8px 16px;
        line-height: 22px;
        margin-left: 0;
      }
    }

    &__image-wrapper {
      padding: 0;
      order: 1;
      margin-bottom: 17px;
      margin-top: 0;
      text-align: center;

      img {
        max-width: 100%;
      }
    }
  }
}
</style>
