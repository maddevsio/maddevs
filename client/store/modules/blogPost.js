import formatDate from '@/helpers/formatDate'

export default {
  state: () => ({
    post: {},
  }),
  mutations: {
    SET_BLOG_POST(state, post) {
      let jsonLd = ''
      let recommendedPosts = []

      if (
        post.data.schema_org_snippets
        && post.data.schema_org_snippets.length
        && post.data.schema_org_snippets[0].single_snippet.length
        && post.data.schema_org_snippets[0].single_snippet[0].text
      ) {
        jsonLd = post.data.schema_org_snippets[0].single_snippet[0].text
        jsonLd = jsonLd.substring(jsonLd.indexOf('{'), jsonLd.lastIndexOf('}') + 1) // extracting only JSON object from a snippet without extra characters
      } else {
        // eslint-disable-next-line
        console.log('Schema.org is not defined');
      }

      if (post.recommendedPosts && post.recommendedPosts.length) {
        recommendedPosts = post.recommendedPosts.filter(recommendedPost => recommendedPost.uid !== post.uid)
        if (recommendedPosts.length > 3) recommendedPosts = recommendedPosts.slice(0, 3)
      }

      state.post = {
        type: post.type,
        id: post.id,
        uid: post.uid,
        title: this.$prismic.asText(post.data.title),
        subtitle: this.$prismic.asText(post.data.subtitle),
        featuredImage: post.data.featured_image,
        postAuthor: post.data.post_author,
        tableOfContents: post.data.table_of_contents,
        slices: post.data.body,
        metaTitle: this.$prismic.asText(post.data.meta_title) || post.data.title[0].text,
        metaDescription: this.$prismic.asText(post.data.meta_description),
        date: formatDate(post.data.date),
        tags: post.tags,
        recommendedPosts,
        jsonLd,
      }
    },
  },
  actions: {
    async getBlogPost({ commit }, payload = { type: 'post' }) {
      try {
        const post = await this.$prismic.api.getByUID(payload.type, payload.uid)

        if (post.tags && post.tags.length) {
          post.recommendedPosts = (
            await this.$prismic.api.query(this.$prismic.predicates.at('document.tags', [post.tags[0]]), {
              pageSize: 4,
            })
          ).results
        }

        commit('SET_BLOG_POST', post)
      } catch (err) {
        if (err) throw err
      }
    },
  },
  getters: {
    blogPost(state) {
      return state.post
    },
  },
}
