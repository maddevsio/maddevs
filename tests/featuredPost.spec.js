import { mount } from '@vue/test-utils'
import FeaturedPost from '../client/components/Blog/FeaturedPost'
import blogPost from './__mocks__/blogPost'

describe('featured post component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(FeaturedPost, {
      propsData: {
        post: blogPost,
      },
      stubs: ['nuxt-link'],
      mocks: {
        $prismic: {
          asText: text => text[0].text,
        },
      },
    })
  })

  it('is a Vue instance', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
