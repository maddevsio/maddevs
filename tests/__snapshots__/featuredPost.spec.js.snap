// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`featured post component renders correctly 1`] = `
<nuxt-link-stub
  class="featured-post"
  to="/blog/its-not-just-a-layout-what-do-you-need-to-know-to/"
>
  <div
    class="row featured-post__wrapper"
  >
    <div
      class="col-12 col-lg-6 featured-post__main"
    >
      <h1
        class="featured-post__title"
      >
        The best practices for improving layout quality on your website
      </h1>
       
      <p
        class="featured-post__paragraph"
      >
        Letâ€™s start with a partial list of how a valid and semantic layout can improve any project:
      </p>
       
      <div
        class="featured-post__data d-flex justify-content-between"
      >
        <div
          class="blog-post__author"
        >
          <img
            alt="Alisa Manakova"
            class="blog-post__author-image"
            src="https://images.prismic.io/superpupertest/255624bf-6a97-48ff-a3b6-337434ba4d33_2_orMuDx-7_VWwpnRL1Q1rxw.jpeg?auto=compress,format"
          />
           
          <div
            class="blog-post__author-info"
          >
            <p
              class="blog-post__author-name"
            >
              Alisa Manakova
            </p>
             
            <div
              class="blog-post__data-of-creation"
            >
              <span
                class="blog-post__author-title"
              >
                Chief Strategy Officer
              </span>
            </div>
          </div>
        </div>
         
        <div
          class="featured-post__meta"
        >
          <div
            class="post-tag dark"
          >
            <span
              class="tag"
            >
              markup
            </span>
          </div>
           
          <span
            class="created-at"
          >
            Aug 25, 2020
          </span>
        </div>
      </div>
    </div>
     
    <div
      class="col-12 col-lg-6 featured-post__image-wrapper"
    >
      <img
        class="featured-post__image"
        height="380"
        src="https://images.prismic.io/superpupertest/e746f4e6-ce33-47b7-b828-5f315fea2bbb_1_bmF_K9TNqtHpHzUUiI_y5Q.jpeg?auto=compress,format"
        width="700"
      />
    </div>
  </div>
</nuxt-link-stub>
`;
